### PR TITLE
DeploymentTarget CR reconciler

### PR DESCRIPTION
**Pull request by jinqi7**
#### Description:
Add DeploymentTarget controller
React to DeploymentTarget CR  with delete action using predicate
Reclaim DeploymentTarget and delete related SpaceRequest CR
Add unit tests for the new functionality

#### Link to JIRA Story (if applicable):

[STONEINTG-336](https://issues.redhat.com/browse/STONEINTG-336)

### CODE REVIEW

File: appstudio-controller/controllers/appstudio.redhat.com/deploymenttargetbinder_controller.go [hunk 1]
```diff
@@ -85,30 +85,57 @@ func (r *DeploymentTargetClaimReconciler) Reconcile(ctx context.Context, req ctr
...
+					log.Info("DeploymentTarget is marked to Deleted", "DeploymentTarget", dt.Name)
+					return ctrl.Result{}, nil
+					//} else if dtcls.Spec.ReclaimPolicy == applicationv1alpha1.ReclaimPolicy_Retain {
+				} else {
+					if removeFinalizer(&dtc, applicationv1alpha1.FinalizerBinder) {
+						if err := r.Client.Update(ctx, &dtc); err != nil {
+							return ctrl.Result{}, fmt.Errorf("failed to remove finalizer %s from DeploymentTargetClaim %s in namespace %s: %v", applicationv1alpha1.FinalizerBinder, dtc.Name, dtc.Namespace, err)
+						}
+						log.Info("Removed finalizer from DeploymentTargetClaim", "finalizer", applicationv1alpha1.FinalizerBinder)
+						dt.Status.Phase = applicationv1alpha1.DeploymentTargetPhase_Released // set phrase to "Released"
```
Comment (dirgim) [inline appstudio-controller/controllers/appstudio.redhat.com/deploymenttargetbinder_controller.go:122]: typo: `set phrase to "Released"` -> `set phase to "Released"`

File: appstudio-controller/controllers/appstudio.redhat.com/deploymenttargetbinder_controller.go [hunk 2]
```diff
@@ -85,30 +85,57 @@ func (r *DeploymentTargetClaimReconciler) Reconcile(ctx context.Context, req ctr
...
 			}
 
 			if dt != nil {
-				err = updateDTStatusPhase(ctx, r.Client, dt, applicationv1alpha1.DeploymentTargetPhase_Released, log)
-				if err != nil {
+				var dtcls *applicationv1alpha1.DeploymentTargetClass
+				if dtcls, err = findMatchingDTClassForDT(ctx, r.Client, *dt); err != nil {
 					return ctrl.Result{}, err
 				}
-				log.Info("DeploymentTarget is released", "DeploymentTarget", dt.Name)
```
Comment (dirgim) [inline appstudio-controller/controllers/appstudio.redhat.com/deploymenttargetbinder_controller.go:100]: Do we want to also log this after DT is set to the `Released` phase?

File: appstudio-controller/controllers/appstudio.redhat.com/deploymenttargetbinder_controller.go [hunk 3]
```diff
@@ -85,30 +85,57 @@ func (r *DeploymentTargetClaimReconciler) Reconcile(ctx context.Context, req ctr
...
+				}
+
+				if dtcls.Spec.ReclaimPolicy == applicationv1alpha1.ReclaimPolicy_Delete {
+					err = r.Client.Delete(ctx, dt)
+					if err != nil {
+						return ctrl.Result{}, err
+					}
+					log.Info("DeploymentTarget is marked to Deleted", "DeploymentTarget", dt.Name)
+					return ctrl.Result{}, nil
+					//} else if dtcls.Spec.ReclaimPolicy == applicationv1alpha1.ReclaimPolicy_Retain {
```
Comment (dirgim) [inline appstudio-controller/controllers/appstudio.redhat.com/deploymenttargetbinder_controller.go:115]: I think we can probably remove this comment for now - if additional reclaim policies get added, that can be handled in a follow-up ticket in the future.

File: appstudio-controller/controllers/appstudio.redhat.com/deploymenttargetreclaimer_controller.go [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+	FinalizerDT = "dt.appstudio.redhat.com/finalizer"
+)
+
+//+kubebuilder:rbac:groups=appstudio.redhat.com,resources=deploymenttargetclaims,verbs=get;list;watch;update;patch
+//+kubebuilder:rbac:groups=appstudio.redhat.com,resources=deploymenttargetclaims/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=appstudio.redhat.com,resources=deploymenttargetclaims/finalizers,verbs=update
+//+kubebuilder:rbac:groups=appstudio.redhat.com,resources=deploymenttargets,verbs=get;list;watch;update;patch;delete
+//+kubebuilder:rbac:groups=appstudio.redhat.com,resources=deploymenttargets/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=appstudio.redhat.com,resources=deploymenttargets/finalizers,verbs=update
+//+kubebuilder:rbac:groups=toolchain.dev.openshift.com,resources=spacerequest,verbs=get;list;watch;create;update;patch;delete
```
Comment (dirgim) [inline appstudio-controller/controllers/appstudio.redhat.com/deploymenttargetreclaimer_controller.go:52]: The resource needs to be plural (`spacerequests`), I had made the same mistake and it had to be fixed in #427

File: appstudio-controller/controllers/appstudio.redhat.com/predicates_dt_test.go [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+
+			It("should ignore when creating a DeploymentTarget", func() {
+				dt.Status.Phase = applicationv1alpha1.DeploymentTargetPhase_Released
+				contextEvent := event.CreateEvent{
+					Object: dt,
+				}
+				Expect(instance.Create(contextEvent)).To(BeFalse())
+			})
+
+			It("should ignore deleting events", func() {
```
Comment (dirgim) [inline appstudio-controller/controllers/appstudio.redhat.com/predicates_dt_test.go:64]: nit: `should ignore deleting events` -> `should pick up on deletion events`

File: appstudio-controller/controllers/appstudio.redhat.com/deploymenttargetreclaimer_controller.go [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+				if err := r.Client.Update(ctx, &dt); err != nil {
+					return ctrl.Result{}, fmt.Errorf("failed to remove finalizer %s from DeploymentTarget %s in namespace %s: %v", FinalizerDT, dt.Name, dt.Namespace, err)
+				}
+				log.Info("Removed finalizer from DeploymentTarget", "finalizer", FinalizerDT)
+				return ctrl.Result{}, nil
+			}
+			return ctrl.Result{}, err
+		}
+		if sr != nil {
+			// delete the ServiceRequest if it has not been deleted
```
Comment (dirgim) [inline appstudio-controller/controllers/appstudio.redhat.com/deploymenttargetreclaimer_controller.go:111]: nit: `ServiceRequest` -> `SpaceRequest`

File: appstudio-controller/controllers/appstudio.redhat.com/deploymenttargetbinder_controller.go (hunk 1 above)
Comment (jinqi7) [reply appstudio-controller/controllers/appstudio.redhat.com/deploymenttargetbinder_controller.go]: Removed it. Thanks!

File: appstudio-controller/controllers/appstudio.redhat.com/deploymenttargetbinder_controller.go (hunk 2 above)
Comment (jinqi7) [reply appstudio-controller/controllers/appstudio.redhat.com/deploymenttargetbinder_controller.go:100]: No, thx

File: appstudio-controller/controllers/appstudio.redhat.com/deploymenttargetbinder_controller.go (hunk 3 above)
Comment (jinqi7) [reply appstudio-controller/controllers/appstudio.redhat.com/deploymenttargetbinder_controller.go]: OK. Thanks!

File: appstudio-controller/controllers/appstudio.redhat.com/deploymenttargetreclaimer_controller.go (hunk 4 above)
Comment (jinqi7) [reply appstudio-controller/controllers/appstudio.redhat.com/deploymenttargetreclaimer_controller.go]: Updated

File: appstudio-controller/controllers/appstudio.redhat.com/predicates_dt_test.go (hunk 5 above)
Comment (jinqi7) [reply appstudio-controller/controllers/appstudio.redhat.com/predicates_dt_test.go]: Updated, thx!

File: appstudio-controller/controllers/appstudio.redhat.com/deploymenttargetreclaimer_controller.go (hunk 6 above)
Comment (jinqi7) [reply appstudio-controller/controllers/appstudio.redhat.com/deploymenttargetreclaimer_controller.go]: Updated, thx!

File: appstudio-controller/controllers/appstudio.redhat.com/deploymenttargetreclaimer_controller.go (hunk 4 above)
Comment (dirgim) [reply appstudio-controller/controllers/appstudio.redhat.com/deploymenttargetreclaimer_controller.go]: Ack, thanks!

File: appstudio-controller/controllers/appstudio.redhat.com/deploymenttargetreclaimer_controller.go (hunk 6 above)
Comment (dirgim) [reply appstudio-controller/controllers/appstudio.redhat.com/deploymenttargetreclaimer_controller.go]: Ack, thanks!

**Review by dirgim — APPROVED** [review]
LGTM, nice work on this one!

**Comment by jinqi7** [comment]
@jgwest , could you please help to review the PR? Thanks

**Review by jgwest — CHANGES_REQUESTED** [review]
Looks great, thanks @jinqi7! Some minor changes, plus test updates requested, then good to go.

**Comment by jgwest** [comment]
@jinqi7 FYI the previous E2E test failed due to this:
```
{"level":"info","ts":"2023-05-03T11:25:24.420078535Z","caller":"runtime/panic.go:838","msg":"Observed a panic in reconciler: runtime error: invalid memory address or nil pointer dereference","controller":"deploymenttargetclaim","controllerGroup":"appstudio.redhat.com","controllerKind":"DeploymentTargetClaim","deploymentTargetClaim":{"name":"test-dtc","namespace":"gitops-service-e2e"},"namespace":"gitops-service-e2e","name":"test-dtc","reconcileID":"56f38c7d-243b-4dc0-a256-07525b755c12"}
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
        panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x118 pc=0x1465952]

goroutine 834 [running]:
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile.func1()
        /go/pkg/mod/github.com/kcp-dev/controller-runtime@v0.12.2-0.20220808200255-4b60fd66e5de/pkg/internal/controller/controller.go:118 +0x1f4
panic({0x15e6360, 0x26302f0})
        /usr/local/go/src/runtime/panic.go:838 +0x207
github.com/redhat-appstudio/managed-gitops/appstudio-controller/controllers/appstudio%2eredhat%2ecom.(*DeploymentTargetClaimReconciler).Reconcile(0xc0007a4408, {0x1a33238, 0xc000b23bf0}, {{{0xc0000d67f8, 0x12}, {0xc000a025f0, 0x8}}, {0x0, 0x0}})
        /workspace/appstudio-controller/controllers/appstudio.redhat.com/deploymenttargetbinder_controller.go:110 +0x8b2
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile(0x1a33190?, {0x1a33238?, 0xc000b23bf0?}, {{{0xc0000d67f8, 0x12}, {0xc000a025f0, 0x8}}, {0x0, 0x0}})
        /go/pkg/mod/github.com/kcp-dev/controller-runtime@v0.12.2-0.20220808200255-4b60fd66e5de/pkg/internal/controller/controller.go:121 +0x122
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler(0xc000220e60, {0x1a33190, 0xc0005fffc0}, {0x166c600?, 0xc000b23ad0?})
        /go/pkg/mod/github.com/kcp-dev/controller-runtime@v0.12.2-0.20220808200255-4b60fd66e5de/pkg/internal/controller/controller.go:320 +0x3b9
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem(0xc000220e60, {0x1a33190, 0xc0005fffc0})
        /go/pkg/mod/github.com/kcp-dev/controller-runtime@v0.12.2-0.20220808200255-4b60fd66e5de/pkg/internal/controller/controller.go:273 +0x1d9
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2()
        /go/pkg/mod/github.com/kcp-dev/controller-runtime@v0.12.2-0.20220808200255-4b60fd66e5de/pkg/internal/controller/controller.go:234 +0x85
created by sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2
        /go/pkg/mod/github.com/kcp-dev/controller-runtime@v0.12.2-0.20220808200255-4b60fd66e5de/pkg/internal/controller/controller.go:230 +0x325
~                                                                                                                                                                                                                                             

```

File: appstudio-controller/controllers/appstudio.redhat.com/deploymenttargetbinder_controller_test.go [hunk 7]
```diff
@@ -116,6 +121,82 @@ var _ = Describe("Test DeploymentTargetClaimBinderController", func() {
 				Expect(dt.Status.Phase).To(Equal(appstudiosharedv1.DeploymentTargetPhase_Released))
 			})
 
+			It("should handle the deletion of a bounded DeploymentTargetClaim", func() {
```
[the remark this answers is not in the document]
Comment (jinqi7) [reply appstudio-controller/controllers/appstudio.redhat.com/deploymenttargetbinder_controller_test.go:124]: If the ReclaimPolicy is "Delete",  the phase/status of DT/DTC will not be changed.  If the ReclaimPolicy is "Retain", the phase of DT will be changed to DeploymentTargetPhase_Released. Added it in that case.

File: appstudio-controller/controllers/appstudio.redhat.com/deploymenttargetbinder_controller.go [hunk 8]
```diff
@@ -85,30 +85,58 @@ func (r *DeploymentTargetClaimReconciler) Reconcile(ctx context.Context, req ctr
...
+
+				if dtcls.Spec.ReclaimPolicy == applicationv1alpha1.ReclaimPolicy_Delete {
+					log.Info("ReclaimPolicy is ReclaimPolicy_Delete")
+					err = r.Client.Delete(ctx, dt)
+					if err != nil {
+						return ctrl.Result{}, err
+					}
+					log.Info("DeploymentTarget is marked to Deleted", "DeploymentTarget", dt.Name)
+					return ctrl.Result{}, nil
+				} else {
```
[the remark this answers is not in the document]
Comment (jinqi7) [reply appstudio-controller/controllers/appstudio.redhat.com/deploymenttargetbinder_controller.go]: Thanks, updated.

File: appstudio-controller/controllers/appstudio.redhat.com/deploymenttargetbinder_controller.go [hunk 9]
```diff
@@ -85,30 +85,58 @@ func (r *DeploymentTargetClaimReconciler) Reconcile(ctx context.Context, req ctr
...
+					log.Info("ReclaimPolicy is ReclaimPolicy_Retain")
```
[the remark this answers is not in the document]
Comment (jinqi7) [reply appstudio-controller/controllers/appstudio.redhat.com/deploymenttargetbinder_controller.go:117]: Thanks, added the test case.

File: appstudio-controller/controllers/appstudio.redhat.com/deploymenttargetreclaimer_controller.go [hunk 10]
```diff
@@ -0,0 +1,201 @@
...
+				readyCond.Reason == "UnableToTerminate" {
+				dt.Status.Phase = applicationv1alpha1.DeploymentTargetPhase_Failed
+				if err := r.Client.Update(ctx, &dt); err != nil {
+					return ctrl.Result{}, err
+				}
+				log.Info("The status of DT %s from %s is updated to Failed", dt.Name, dt.Namespace)
+				return ctrl.Result{}, nil
+			}
+
+			return ctrl.Result{}, nil
```
[the remark this answers is not in the document]
Comment (jinqi7) [reply appstudio-controller/controllers/appstudio.redhat.com/deploymenttargetreclaimer_controller.go]: Thanks, updated.

File: appstudio-controller/controllers/appstudio.redhat.com/deploymenttargetreclaimer_controller.go [hunk 11]
```diff
@@ -0,0 +1,201 @@
...
+			if removeFinalizer(&dt, FinalizerDT) {
+				if err := r.Client.Update(ctx, &dt); err != nil {
+					return ctrl.Result{}, fmt.Errorf("failed to remove finalizer %s from DeploymentTarget %s in namespace %s: %v", FinalizerDT, dt.Name, dt.Namespace, err)
+				}
+				log.Info("Removed finalizer from DeploymentTarget", "finalizer", FinalizerDT)
+				return ctrl.Result{}, nil
+			}
+			return ctrl.Result{}, err
+		}
+		if sr != nil {
```
[the remark this answers is not in the document]
Comment (jinqi7) [reply appstudio-controller/controllers/appstudio.redhat.com/deploymenttargetreclaimer_controller.go:110]: I added the finalizer for ServiceRequest before Delete it although  host-operator also check and add it. Then, we can check the block of deletion failed. Please help to review, Thanks!

File: appstudio-controller/controllers/appstudio.redhat.com/predicates_dt.go [hunk 12]
```diff
@@ -0,0 +1,37 @@
...
+		},
+		DeleteFunc: func(deleteEvent event.DeleteEvent) bool {
+			return true
+		},
+		GenericFunc: func(genericEvent event.GenericEvent) bool {
+			return false
+		},
+		UpdateFunc: func(e event.UpdateEvent) bool {
+			return hasDTDeletionTimestampChanged(e.ObjectOld, e.ObjectNew)
+		},
```
[the remark this answers is not in the document]
Comment (jinqi7) [reply appstudio-controller/controllers/appstudio.redhat.com/predicates_dt.go:26]: Currently, in both delete and update events, when the  DeletionTimestamp is detected to be existing from DT CR, the reconcile will be triggered. We are not relying on the change of  the deletion timestamp.

File: appstudio-controller/controllers/appstudio.redhat.com/deploymenttargetbinder_controller.go [hunk 13]
```diff
@@ -85,30 +85,58 @@ func (r *DeploymentTargetClaimReconciler) Reconcile(ctx context.Context, req ctr
...
+					if removeFinalizer(&dtc, applicationv1alpha1.FinalizerBinder) {
+						if err := r.Client.Update(ctx, &dtc); err != nil {
+							return ctrl.Result{}, fmt.Errorf("failed to remove finalizer %s from DeploymentTargetClaim %s in namespace %s: %v", applicationv1alpha1.FinalizerBinder, dtc.Name, dtc.Namespace, err)
+						}
+						log.Info("Removed finalizer from DeploymentTargetClaim", "finalizer", applicationv1alpha1.FinalizerBinder)
+						err := updateDTStatusPhase(ctx, r.Client, dt, applicationv1alpha1.DeploymentTargetPhase_Released, log)
+						if err != nil {
+							return ctrl.Result{}, fmt.Errorf("failed to update DeploymentTarget %s in namespace %s to Released status", dt.Name, dt.Namespace)
+						}
+						return ctrl.Result{}, nil
```
[the remark this answers is not in the document]
Comment (jinqi7) [reply appstudio-controller/controllers/appstudio.redhat.com/deploymenttargetbinder_controller.go:127]: It's to implement that "The DeploymentTarget controller should see that the DTClaim was deleted, and the DeploymentTarget’s .status.phase field should be set to Released".

File: appstudio-controller/controllers/appstudio.redhat.com/deploymenttargetbinder_controller.go [hunk 14]
```diff
@@ -85,30 +85,58 @@ func (r *DeploymentTargetClaimReconciler) Reconcile(ctx context.Context, req ctr
...
 		}
-
 		if removeFinalizer(&dtc, applicationv1alpha1.FinalizerBinder) {
 			if err := r.Client.Update(ctx, &dtc); err != nil {
 				return ctrl.Result{}, fmt.Errorf("failed to remove finalizer %s from DeploymentTargetClaim %s in namespace %s: %v", applicationv1alpha1.FinalizerBinder, dtc.Name, dtc.Namespace, err)
 			}
 			log.Info("Removed finalizer from DeploymentTargetClaim", "finalizer", applicationv1alpha1.FinalizerBinder)
+			return ctrl.Result{}, nil
 		}
-
```
[the remark this answers is not in the document]
Comment (jinqi7) [reply appstudio-controller/controllers/appstudio.redhat.com/deploymenttargetbinder_controller.go:110]: It's to handle the deletion of a bounded DeploymentTargetClaim with deleted DeploymentTarget which is a test case in the unit test.

**Comment by jinqi7** [comment]
For the e2e-tests, I reran it and didn't reproduce the failure mentioned in the two comments. @jgwest

**Comment by jgwest** [comment]
> For the e2e-tests, I reran it and didn't reproduce the failure mentioned in the two comments. @jgwest

Thanks for checking @jinqi7! I was able to reproduce, and pushed a fix to resolve the issue. The failure rate was very intermittent, only reproducing in about ~10% of runs.

**Review by jgwest — APPROVED** [review]
Looks great, thanks again @jinqi7!

**Pull request merged**